### PR TITLE
Feature/468 basic landing page

### PR DESCRIPTION
**Pull request by lokm01**

### CODE REVIEW

File: menas/src/main/scala/za/co/absa/enceladus/menas/controllers/LandingPageController.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.security.core.context.SecurityContextHolder
+import org.springframework.security.core.userdetails.UserDetails
+import org.springframework.web.bind.annotation._
+import za.co.absa.enceladus.model.user.UserInfo
+import za.co.absa.enceladus.menas.models.LandingPageInformation
+import za.co.absa.enceladus.menas.repositories.DatasetMongoRepository
+import za.co.absa.enceladus.menas.repositories.MappingTableMongoRepository
+import za.co.absa.enceladus.menas.repositories.SchemaMongoRepository
+import java.util.concurrent.CompletableFuture
+import org.springframework.beans.factory.annotation.Autowired
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/controllers/LandingPageController.scala]: Unused imports on line 18, 19, 20, 21 and 23

File: menas/src/main/scala/za/co/absa/enceladus/menas/controllers/LandingPageController.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+@RestController
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/controllers/LandingPageController.scala]: trailing whitespace

File: menas/src/main/scala/za/co/absa/enceladus/menas/models/LandingPageInformation.scala [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.menas.models
+
+case class DistinctCount(distinctCount: Int)
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/models/LandingPageInformation.scala]: trailing whitespace

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala [hunk 4]
```diff
@@ -41,12 +42,19 @@ abstract class VersionedMongoRepository[C <: VersionedModel](mongoDb: MongoDatab
   private def getParent(oldEntity: C): MenasReference = {
     MenasReference(collection = Some(collectionBaseName), name = oldEntity.name, version = oldEntity.version)
   }
-  
+
+  def distinctCount(): Future[Int] = {
+    //    collection.distinct("name", getNotDisabledFilter).
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala]: residual line

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala [hunk 5]
```diff
@@ -55,42 +63,38 @@ abstract class VersionedMongoRepository[C <: VersionedModel](mongoDb: MongoDatab
...
   }
 
   def update(username: String, updated: C): Future[C] = {
     for {
       latestVersion <- getLatestVersionValue(updated.name)
-      newVersion <- if(latestVersion.isEmpty) throw new NotFoundException()
-           else if(latestVersion.get != updated.version) throw new EntityAlreadyExistsException(s"Entity ${updated.name} (version. ${updated.version}) already exists.") 
-           else Future.successful(latestVersion.get + 1)
+      newVersion <- if (latestVersion.isEmpty) throw new NotFoundException()
+      else if (latestVersion.get != updated.version) throw new EntityAlreadyExistsException(s"Entity ${updated.name} (version. ${updated.version}) already exists.")
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala]: Please add some braces and new lines. + the exceptions are case classes so `new` is not needed

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala [hunk 6]
```diff
@@ -55,42 +63,38 @@ abstract class VersionedMongoRepository[C <: VersionedModel](mongoDb: MongoDatab
...
       res <- collection.insertOne(newInfo).toFuture()
     } yield newInfo
-    
+
   }
 
   def disableVersion(name: String, version: Option[Int], username: String): Future[UpdateResult] = {
-    collection.updateMany(getNameVersionFilter(name, version), combine(
-      set("disabled", true),
+    collection.updateMany(getNameVersionFilter(name, version), combine(set("disabled", true),
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala]: Why was this moved up?

File: menas/ui/components/home/landingPage.controller.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+   * modify the View before it is displayed, to bind event handlers and do other one-time initialization.
+   * 
+   * @memberOf components.home.landingPage
+   */
+  onInit: function() {
+    this._router = sap.ui.core.UIComponent.getRouterFor(this);
+    this._router.getRoute("home").attachMatched(function(oEvent) {
+      const config = oEvent.getParameter("config");
+      this._appId = `${config.targetParent}--${config.controlId}`;
+      this._app = sap.ui.getCore().byId(this._appId)
```
Comment (Zejnilovic) [inline menas/ui/components/home/landingPage.controller.js]:
```suggestion
      this._app = sap.ui.getCore().byId(this._appId);
```

File: menas/ui/components/home/landingPage.controller.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    if(sTarget === "datasets") {
+      viewBase = `${viewBase}--datasetsPage`;
+    } else if(sTarget === "schemas") {
+      viewBase = `${viewBase}--schemasPage`;
+    } else if(sTarget === "mappingTables") {
+      viewBase = `${viewBase}--mappingTablesPage`;
+    }
+    this._eventBus.publish(sTarget, "list");
+
+    this._app.backToTopMaster()
```
Comment (Zejnilovic) [inline menas/ui/components/home/landingPage.controller.js]:
```suggestion
    this._app.backToTopMaster();
```

File: menas/ui/components/home/landingPage.controller.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    this._app.toMaster(viewBase);
+    setTimeout((() => {this._app.showMaster()}).bind(this), 300)
+  }
+/**
+ * Similar to onAfterRendering, but this hook is invoked before the controller's View is re-rendered (NOT before the
+ * first rendering! onInit() is used for that one!).
+ * 
+ * @memberOf components.home.landingPage
+ */
```
Comment (Zejnilovic) [inline menas/ui/components/home/landingPage.controller.js]: Is this in preparation for the future?

File: menas/ui/components/home/landingPage.controller.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    } else if(sTarget === "schemas") {
+      viewBase = `${viewBase}--schemasPage`;
+    } else if(sTarget === "mappingTables") {
+      viewBase = `${viewBase}--mappingTablesPage`;
+    }
+    this._eventBus.publish(sTarget, "list");
+
+    this._app.backToTopMaster()
+    this._app.toMaster(viewBase);
+    setTimeout((() => {this._app.showMaster()}).bind(this), 300)
```
Comment (Zejnilovic) [inline menas/ui/components/home/landingPage.controller.js]: Maybe we could simplify this if it is meant to be a oneliner.
```suggestion
    setTimeout(this._app.showMaster().bind(this), 300)
```

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala (hunk 6 above)
Comment (GeorgiChochov) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala]: Everything is getting inlined, I guess that's an editor configuration.

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala [hunk 11]
```diff
@@ -41,12 +42,19 @@ abstract class VersionedMongoRepository[C <: VersionedModel](mongoDb: MongoDatab
   private def getParent(oldEntity: C): MenasReference = {
     MenasReference(collection = Some(collectionBaseName), name = oldEntity.name, version = oldEntity.version)
   }
-  
+
+  def distinctCount(): Future[Int] = {
```
Comment (GeorgiChochov) [inline menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala]: How about some tests (at least add a couple case to `DatasetRepositoryIntegrationSuite`)?

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala (hunk 6 above)
Comment (lokm01) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala]: Yeah autoformat.

File: menas/ui/components/home/landingPage.controller.js (hunk 10 above)
Comment (lokm01) [reply menas/ui/components/home/landingPage.controller.js]: Yeah I'll simplify

File: menas/ui/components/home/landingPage.controller.js (hunk 9 above)
Comment (lokm01) [reply menas/ui/components/home/landingPage.controller.js]: UI5 IDE template :)

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala [hunk 12]
```diff
@@ -41,12 +42,19 @@ abstract class VersionedMongoRepository[C <: VersionedModel](mongoDb: MongoDatab
...
+    //    collection.distinct("name", getNotDisabledFilter).
+    val pipeline = Seq(filter(getNotDisabledFilter),
+      Aggregates.group("$name", Accumulators.max("max", "$version")),
```
[the remark this answers is not in the document]
Comment (lokm01) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/VersionedMongoRepository.scala]: Yep thanks!

File: menas/src/main/scala/za/co/absa/enceladus/menas/models/LandingPageInformation.scala [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.menas.models
+
+case class DistinctCount(distinctCount: Int)  
+
+case class LandingPageInformation(
```
[the remark this answers is not in the document]
Comment (lokm01) [reply menas/src/main/scala/za/co/absa/enceladus/menas/models/LandingPageInformation.scala]: Ok agree with DistinctCount.

Landing page will eventually contain much more information.
